store/postgres/src/store_events.rs

```diff
@@ -0,0 +1,141 @@
+use diesel::deserialize::QueryableByName;
+use diesel::pg::Pg;
+use diesel::pg::PgConnection;
+use diesel::sql_types::Text;
+use diesel::RunQueryDsl;
+use graph::prelude::*;
+use graph::serde_json;
+use notification_listener::{NotificationListener, SafeChannelName};
+
+pub struct StoreEventListener {
+    notification_listener: NotificationListener,
+}
+
+impl StoreEventListener {
+    pub fn new(postgres_url: String) -> Self {
+        StoreEventListener {
+            notification_listener: NotificationListener::new(
+                postgres_url,
+                SafeChannelName::i_promise_this_is_safe("store_events"),
+            ),
+        }
+    }
+
+    pub fn start(&mut self) {
+        self.notification_listener.start()
+    }
+}
+
+impl EventProducer<StoreEvent> for StoreEventListener {
+    fn take_event_stream(&mut self) -> Option<StoreEventStreamBox> {
+        self.notification_listener.take_event_stream().map(
+            |stream| -> Box<Stream<Item = _, Error = _> + Send> {
+                Box::new(stream.map(|notification| {
+                    // Parse notification as JSON
+                    let value: serde_json::Value = serde_json::from_str(&notification.payload)
+                        .expect(&format!(
+                            "invalid JSON store event received from database: '{}'",
+                            &notification.payload
+                        ));
+
+                    // Create StoreEvent from JSON
+                    let change: StoreEvent =
+                        serde_json::from_value(value.clone()).unwrap_or_else(|_| {
+                            panic!("invalid store event received from database: {:?}", value)
+                        });
+
+                    change
+                }))
+            },
+        )
+    }
+}
+
+// Querying the entity changes corresponding to the reversion of a certain
+// block for a certain subgraph. We want to keep the deserialization
+// logic (QueryableByName) close to the actual query. That's why we introduce
+// the intermediate EntityChangeQBN struct
+
+struct EntityChangeQBN(EntityChange);
+
+impl QueryableByName<Pg> for EntityChangeQBN {
+    fn build<R: diesel::row::NamedRow<diesel::pg::Pg>>(
+        row: &R,
+    ) -> diesel::deserialize::Result<Self> {
+        let subgraph_id: String = row.get("subgraph")?;
+        let subgraph_id = SubgraphDeploymentId::new(subgraph_id).map_err(|_| {
+            format!(
+                "Illegal subgraph_id {}",
+                row.get("subgraph").unwrap_or("<unknown>".to_string())
```

Comment:
Since we've already obtained the subgraph ID in line 65, let's just use that here. May have to clone it to make it available in the closure but that's ok.

Reply:
An alternative would be to do `row.get("subgraph").unwrap()` here, since we would've returned an error in line 65 already if anything was wrong with the `"subgraph"` entry in the row.

Reply:
I'd rather go with the second alternative, since using the existing `subgraph_id` requires that we clone for every result row, even though this whole error should never happen. (Though, to be sure, we'd need to put a check constraint on that column, so that we can just unwrap the `SubgraphDeploymentId`)

Reply:
You can't clone inside the closure so it's only cloned if necessary?

Reply:
No, the `SubgraphDeploymentId::new(subgraph_id)` takes ownership of the value.

Reply:
Ah I see. Well subgraph ids have a fixed lenght, so you can reliably clone them in nanoseconds, I'd only worry about cloning variable length things.